### PR TITLE
Factorio: Added ability to chat from within the game.

### PR DESCRIPTION
**Pull request by recklesscoder**
## What is this fixing or adding?
So far it was possible to receive Archipelago messages and chat within the Factorio game client, but chat messages written within the game would not be bridged to Archipelago. This also meant Archipelago commands such as `!hint` could not be used from within the game.

This PR rectifies the deficiency described above. I mimicked the Minecraft implementation: sent chat messages are prepended the Factorio player name in parentheses, except when an AP command is sent (anything starting with `"!"`).

## How was this tested?
I tested this locally, sending messages between a Factorio instance and a Minecraft instance via Archipelago.
Commands work fine. Chat message echoes are caught, even when the Archipelago player has an alias.

## Potential Issues
The full output of `!help` is a little too tall for my screen at 1080p and 100% UI scaling. Unfortunately, Factorio does not implement a scrolling mechanism for its console.
A similar (and common) problem may occur when the Factorio AP world option `tech_tree_information: full` is set, bestowing nearly 200 hints upon the player, such that the output of `!hint` is far too long as well. As new and specifically queried hints appear without the full list and the available hint points appear below the list, the hinting feature is fully usable, just spammy. The Factorio AP client window suffers from the same spam, although it at least allows you to scroll up if you dare.

Overall I still deem the addition of this feature an improvement.

### CODE REVIEW

**Comment by recklesscoder** [comment]
Thinking about this addition some more, I wonder whether you (whoever is reading this) think these features would be desirable:
- Not bridging messages consisting solely or partially of local coordinates/"pings" like `[gps=81.9,-56.7]`
- Configuration options (presumably living in `host.yaml`) for
  - Enabling/disabling bridging out chat from within the game (for when Factorio players want to communicate without broadcasting to Archipelago)
  - Enabling/disabling prepending the Factorio player name to bridged chat messages (as it is redundant if there is only one person playing on the server)
- FactorioClient and/or in-game commands for changing these options at run time

Let me know!

**Comment by Berserker66** [comment]
I agree with the proposed improvements. The username prepending could be automatic in case of playercount > 1 in the savegame.

**Comment by recklesscoder** [comment]
I like that solution, I'll look into it.

I'm still looking for feedback on which of the other additional features I proposed in my previous comment I should implement (unless you were referring to all of them). I'm not sure how acceptable it is to just willy-nilly add options to `host.yaml`.

**Comment by Berserker66** [comment]
that's kind of what the file is for, it's the dumping ground for options that have nowhere else to go. and yeah I was referring to the bunch.

**Comment by recklesscoder** [comment]
Naming the commands is an awkward balancing act between brevity and accuracy. I went with `/toggle_chat` in FactorioClient and `/toggle-ap-chat` in the Factorio client for now. All the other in-game commands introduced by Archipelago are prefixed with `ap-` but are not meant to be used by players directly, so I tried to convey the difference by breaking with that naming convention.

I did not update the setup guide with an explanation of the new host.yaml setting, nor the client toggle command, nor the in-game toggle command, as I am unsure whether that is in scope. Turning the chat bridge off seems like a niche feature to me. It being documented in the respective `/help` commands and the default host.yaml might be enough.

File: FactorioClient.py [hunk 1]
```diff
@@ -123,6 +132,38 @@ def on_package(self, cmd: str, args: dict):
...
+        if not self.bridge_chat_out:
+            return
+
+        # Pass through commands
+        if message.startswith("!"):
+            await self.send_msgs([{"cmd": "Say", "text": message}])
+            return
+
+        # Remove local coordinates
+        message = re.sub(r"\[gps=[^\]]*\]", "", message).strip()
```
Comment (black-sliver) [inline FactorioClient.py:150]: if we assume all messages with `[gps=` are pings and should not get sent, we can do an easier, faster and more readable
`if "[gps=" in message: return`

File: worlds/factorio/data/mod_template/control.lua [hunk 2]
```diff
@@ -157,6 +157,7 @@ function on_player_created(event)
 {%- if silo == 2 %}
     check_spawn_silo(game.players[event.player_index].force)
 {%- endif %}
+    dumpInfo(player.force)
```
Comment (black-sliver) [inline worlds/factorio/data/mod_template/control.lua:160]: i don't see this being used?

File: FactorioClient.py (hunk 1 above)
Comment (recklesscoder) [reply FactorioClient.py:150]: If you wish to assume that, we can do that, sure. Note that it is also possible to have a mixed chat message by pinging with an open chat box. One might write "i have a lot of iron over here [gps=4711,815]".

File: worlds/factorio/data/mod_template/control.lua (hunk 2 above)
Comment (recklesscoder) [reply worlds/factorio/data/mod_template/control.lua:160]: Currently it is not being used, as `dumpInfo` in general is currently not used. Instead we are currently polling once a second regardless of need. However, as there seems to be a desire to eventually go back to the "only poll when necessary" system (see one of the TODOs in FactorioClient.py), I thought I'd make this feature compatible with that other system as well. (The `dumpInfo` in `on_player_created` specifically would then be required to update the `FactorioContext.multiplayer` flag when a second player joins.)

File: FactorioClient.py (hunk 1 above)
Comment (black-sliver) [reply FactorioClient.py:150]: yeah, my thinking is that a ping should never concern players outside the factorio session

Comment (recklesscoder) [reply FactorioClient.py:150]: My thinking was that removing the entire mixed message could lead to confusion when replies to such messages are bridged.

Comment (black-sliver) [reply FactorioClient.py:150]: i think `see iron here [gps=]` is the most common scenario, which is worse than having replies you don't understand

**Review by black-sliver — APPROVED** [review]
Thanks!

**Pull request merged**